### ISSUE TITLE
Fix ClassCastException in JS Enrichment

### ISSUE THREAD

**Issue by chuwy** [snowplow/beam-enrich#42]
Source: https://github.com/snowplow/snowplow/issues/4349

The first step would be to port a missing specification for JS Enrichment from SCE: https://github.com/snowplow/beam-enrich/tree/master/src/test/scala/com.snowplowanalytics.snowplow.enrich.beam/enrichments

**Comment by jiristepan** [comment]
I did aditional test with javascript_script_enricher.json

```
{
    "schema": "iglu:com.snowplowanalytics.snowplow/javascript_script_config/jsonschema/1-0-0",
    "data": {
        "vendor": "com.snowplowanalytics.snowplow",
        "name": "javascript_script_config",
        "enabled": true,
        "parameters": {
            "script": "ZnVuY3Rpb24gcHJvY2VzcyhldmVudCkgewogICAgcmV0dXJuIFtdOwp9"
        }
    }
}
```

It contains minimalistics proces() functions:
```
function process(event) {
    return [];
}
```

The result is the same:
```
[main] INFO org.apache.beam.runners.dataflow.DataflowPipelineTranslator - Adding raw-from-pubsub/MapElements/Map as step s2
Exception in thread "main" java.lang.ClassCastException: cannot assign instance of java.lang.invoke.SerializedLambda to field com.spotify.scio.util.Functions$$anon$7.g of type scala.Function1 in instance of com.spotify.scio.util.Functions$$anon$7
        at java.io.ObjectStreamClass$FieldReflector.setObjFieldValues(ObjectStreamClass.java:2301)
        at java.io.ObjectStreamClass.setObjFieldValues(ObjectStreamClass.java:1431)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2350)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2268)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2126)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1625)
        at java.io.ObjectInputStream.defaultReadFields(ObjectInputStream.java:2344)
        at java.io.ObjectInputStream.readSerialData(ObjectInputStream.java:2268)
        at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:2126)
        at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1625)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:465)
        at java.io.ObjectInputStream.readObject(ObjectInputStream.java:423)
        at org.apache.beam.sdk.util.SerializableUtils.deserializeFromByteArray(SerializableUtils.java:71)
        at org.apache.beam.runners.core.construction.ParDoTranslation.doFnWithExecutionInformationFromProto(ParDoTranslation.java:610)
        at org.apache.beam.runners.core.construction.ParDoTranslation.getSchemaInformation(ParDoTranslation.java:314)
        at org.apache.beam.runners.core.construction.ParDoTranslation.getSchemaInformation(ParDoTranslation.java:299)
        at org.apache.beam.runners.dataflow.DataflowPipelineTranslator$8.translateSingleHelper(DataflowPipelineTranslator.java:1003)
        at org.apache.beam.runners.dataflow.DataflowPipelineTranslator$8.translate(DataflowPipelineTranslator.java:995)
        at org.apache.beam.runners.dataflow.DataflowPipelineTranslator$8.translate(DataflowPipelineTranslator.java:992)
        at org.apache.beam.runners.dataflow.DataflowPipelineTranslator$Translator.visitPrimitiveTransform(DataflowPipelineTranslator.java:494)
        at org.apache.beam.sdk.runners.TransformHierarchy$Node.visit(TransformHierarchy.java:665)
        at org.apache.beam.sdk.runners.TransformHierarchy$Node.visit(TransformHierarchy.java:657)
        at org.apache.beam.sdk.runners.TransformHierarchy$Node.access$600(TransformHierarchy.java:317)
        at org.apache.beam.sdk.runners.TransformHierarchy.visit(TransformHierarchy.java:251)
        at org.apache.beam.sdk.Pipeline.traverseTopologically(Pipeline.java:460)
        at org.apache.beam.runners.dataflow.DataflowPipelineTranslator$Translator.translate(DataflowPipelineTranslator.java:433)
        at org.apache.beam.runners.dataflow.DataflowPipelineTranslator.translate(DataflowPipelineTranslator.java:192)
        at org.apache.beam.runners.dataflow.DataflowRunner.run(DataflowRunner.java:797)
        at org.apache.beam.runners.dataflow.DataflowRunner.run(DataflowRunner.java:188)
        at org.apache.beam.sdk.Pipeline.run(Pipeline.java:315)
        at org.apache.beam.sdk.Pipeline.run(Pipeline.java:301)
        at com.spotify.scio.ScioContext.execute(ScioContext.scala:598)
        at com.spotify.scio.ScioContext$$anonfun$run$1.apply(ScioContext.scala:586)
        at com.spotify.scio.ScioContext$$anonfun$run$1.apply(ScioContext.scala:574)
        at com.spotify.scio.ScioContext.requireNotClosed(ScioContext.scala:694)
        at com.spotify.scio.ScioContext.run(ScioContext.scala:574)
        at com.snowplowanalytics.snowplow.enrich.beam.Enrich$.main(Enrich.scala:91)
        at com.snowplowanalytics.snowplow.enrich.beam.Enrich.main(Enrich.scala)
```

**Comment by chuwy** [comment]
Might be related to https://github.com/spotify/scio/issues/2778

**Comment by chuwy** [comment]
Also seems @peel already discovered that (https://github.com/snowplow/beam-enrich/issues/20), but assumed it happens only for invalid JS snippet.

**Comment by benjben** [comment]
This bug is there since at least Beam Enrich 1.0.0. I didn't try with 0.4.0, as the integration tests back then were in Stream Enrich.

**Comment by benjben** [comment]
The issue seems to be that the javascript function provided in the enrichment configuration is compiled into a `SerializedLambda` and passed to the workers, but the workers are missing the class which defined the lambda expression and can't deserialize it.

Using [Nashorn](https://www.oracle.com/technical-resources/articles/java/jf14-nashorn.html) instead of Rhino might seem the incompatibility issue:

> Until Java SE 7, JDKs shipped with a JavaScript scripting engine based on Mozilla Rhino. Java SE 8 will instead ship with a new engine called Oracle Nashorn

**Comment by rbkn** [comment]
Unsure if related to this issue (using SCIO) but Google implements js transforms in some of its predefined Java Dataflow templates.

e.g.
[TextToBigQueryStreaming.java](https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/master/src/main/java/com/google/cloud/teleport/templates/TextToBigQueryStreaming.java#L193)

Might help in checking what will be compatible with Dataflow Runner in the future.

**Comment by chuwy** [comment]
Closing in favor of https://github.com/snowplow/enrich/issues/138. This should be fixed in 1.3.0-rc6 (although we haven't tested it yet).
